utils/src/main/java/software/amazon/awssdk/utils/async/AsyncBufferingSubscriber.java

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.utils.async;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Function;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import software.amazon.awssdk.annotations.SdkProtectedApi;
+import software.amazon.awssdk.utils.Logger;
+import software.amazon.awssdk.utils.Validate;
+
+/**
+ * An implementation of {@link Subscriber} that delivers data to async consumer and limits the number of concurrent requests by
+ * only requesting data if the number of inflight request is less than the configured maxConcurrentRequests.
+ *
+ * @param <T> Type of data requested
+ */
+@SdkProtectedApi
+public class AsyncBufferingSubscriber<T> implements Subscriber<T> {
+    private static final Logger log = Logger.loggerFor(AsyncBufferingSubscriber.class);
+    private final Event<T> completeEvent = new CompleteEvent<>();
+    private final Queue<Event<T>> buffer;
+    private final CompletableFuture<?> future;
+    private final Function<T, CompletableFuture<?>> consumer;
+    private final int maxConcurrentRequests;
+    private final AtomicInteger numRequestsInFlight;
+    private final AtomicBoolean isDelivering = new AtomicBoolean(false);
+    private volatile boolean isStreamingDone;
+    private Subscription subscription;
+    private final List<CompletableFuture<?>> futures;
+
+    public AsyncBufferingSubscriber(Function<T, CompletableFuture<?>> consumer,
+                                    CompletableFuture<Void> future,
+                                    int maxConcurrentRequests) {
+        this.buffer = new ConcurrentLinkedQueue<>();
+        this.future = future;
+        this.consumer = consumer;
+        this.maxConcurrentRequests = maxConcurrentRequests;
+        this.numRequestsInFlight = new AtomicInteger(0);
```

Comment:
We don't provide synchronization to guarantee that `buffer` and `numRequestsInFlight` are updated atomically. Can this result in any race conditions with the current logic?

Reply:
Thanks, I confused myself. So we have `numRequestsInFlight` and `maxConcurrentRequests`, and we only request more objects when `cur` < `max`, right? Won't this result in the queue/buffer being almost always empty?

Should we have an additional parameter, e.g., `maxBacklog`, that dictates the max the queue is allowed to grow? It seems like an easy addition since the queue logic already exists. Otherwise, I'm not if we even really need a queue at all.

Reply:
The Javadoc I wrote is wrong...I'll update it.

Reply:
To close the loop, we discussed offline and decided to defer this and keep this class as an internal API in s3-transfer-manager module for now.